yarn-project/noir-contracts/src/contracts/native_token_contract/src/main.nr

```diff
@@ -0,0 +1,494 @@
+mod storage;
+
+// Testing token that can be bridged in and out.
+// TODOS:
+// - Add role based access control to mint functions
+// - Add function for managing roles
+// - Add public self-burn function for users to burn their own tokens
+contract NativeToken {
+    // Libs
+    use dep::value_note::{
+        balance_utils,
+        utils::{send_note, spend_notes},
+        value_note::{VALUE_NOTE_LEN, ValueNoteMethods},
+    };
+
+    use dep::non_native::{
+        hash::{get_mint_content_hash, get_withdraw_content_hash},
+        transparent_note::{
+            TransparentNote,
+            TransparentNoteMethods
+        },
+    };
+
+    use crate::storage::Storage;
+
+    use dep::aztec::context::{
+        PrivateContext,
+        PublicContext,
+    };
+    use dep::aztec::types::point::Point;
+    use dep::aztec::{
+        abi,
+        abi::{
+            PublicContextInputs,
+            PrivateContextInputs,
+        },
+        note::{
+            note_header::NoteHeader,
+            utils as note_utils,
+        },
+    };
+
+    // oracles
+    use dep::aztec::oracle::{
+        create_l2_to_l1_message::create_l2_to_l1_message,
+    };
+
+    // public messaging
+    use dep::aztec::messaging::consume_l1_to_l2_message_public;
+    use dep::aztec::public_call_stack_item::PublicCallStackItem;
+
+    use dep::aztec::oracle::create_commitment::create_commitment;
+
+    fn constructor(
+        //*********************************/
+        // Should eventually be hidden:
+        inputs: PrivateContextInputs,
+        //*********************************/
+        initial_supply: Field,
+        owner: Field,
+    ) -> distinct pub abi::PrivateCircuitPublicInputs {
+        let storage = Storage::init();
+
+        let mut context = PrivateContext::new(inputs, abi::hash_args([initial_supply, owner]));
+
+        let balance = storage.balances.at(owner);
+        send_note(&mut context, balance, initial_supply, owner);
+
+        context.finish()
+    }
+
+    // uint256(keccak256("owner_mint_pub(field,field)")) >> 224 -> 1071038680
+    open fn owner_mint_pub(
+        inputs: PublicContextInputs,
+        to: Field,
+        amount: Field,
+    ) -> pub abi::PublicCircuitPublicInputs {
+        let mut context = PublicContext::new(inputs, abi::hash_args([to, amount]));
+        let storage = Storage::init();
+        let new_balance = storage.public_balances.at(to).read() + amount;
+        storage.public_balances.at(to).write(new_balance);
+        storage.total_supply.write(storage.total_supply.read() + amount);
+        context.return_values.push(1);
```

Comment:
What's this for? Is it similar to how `transferFrom()` returns `bool`?

Reply:
At some point there was failures if no return values, so leftover from that.